### PR TITLE
Improved CSS tooling

### PR DESCRIPTION
**Pull request by fverloop**
I've added a couple of things that will help maintain and write consistent CSS. I've used the following packages:

- [Stylelint](https://github.com/stylelint/stylelint)
- [Colorguard](https://github.com/SlexAxton/css-colorguard)
- [Autoprefixer](https://github.com/postcss/autoprefixer)
- [Sourcemaps](https://www.npmjs.com/package/gulp-sourcemaps)

---

### Stylelint
To keep CSS consistent throughout the project, styles are now checked and validated against a set of rules. Every project can have a custom configuration. The [rules](http://stylelint.io/user-guide/rules/) are defined in a `.stylelintrc` file that lives in the root of a Moonbase project. An alternative option is adding the rules to a `stylelint` property in `package.json`, but I feel this is less obvious. Every new Moonbase project should start with default configuration. 

For our Framer site I've set up a config file based on the following rules:
- Use hard-tabs
- Put spaces after `:` in property declarations.
- Put spaces before `{` in rule declarations.
- Blocks can't be on a single line
- When grouping selectors, keep individual selectors to a single line.
- Place closing braces of declaration blocks on a new line.
- Each declaration should appear on its own line for more accurate error reporting.
- Comments should have an empty line before
- Color hex codes should be written in uppercase
- There can't be more then 2 empty lines
- Use double quotes

Besides that it checks for:
- Duplicate selectors
- Duplicate properties
- Empty blocks
- Invalid hex codes

The config file looks like this:
```JSON
{
	"rules": {
		"block-no-empty": true,
		"color-no-invalid-hex": true,
		"comment-empty-line-before": [ "always", {
			"ignore": ["stylelint-commands", "between-comments"],
		} ],
		"declaration-colon-space-after": "always",
		"indentation": ["tab", {
			"except": ["value"]
		}],
		"max-empty-lines": 2,
		"color-hex-case": "upper",
		"string-quotes": "double",
		"declaration-block-no-duplicate-properties": true,
		"declaration-block-trailing-semicolon": "always",
		"block-opening-brace-space-before": "always",
		"declaration-block-semicolon-newline-after": "always",
		"block-closing-brace-newline-after": "always",
		"block-no-single-line": true,
		"selector-list-comma-newline-after": "always",
		"no-duplicate-selectors": true
	},
	"ignoreFiles": ["./assets/css/docs-style.scss", "assets/css/components/reset.scss"],
	"defaultSeverity": "warning"
}
```

### Reporting
I've updated the reporting task to strip out all common reset classes. It warns you when colors you've added are too similar to ones that already exist. The output format is still quite ugly here.

### SCSS build task
This one is pretty straight forward. I've added Autoprefixer, so all properties will be automatically prefixed when needed. And I've enabled sourcemaps so debugging will be a bit better.

### CODE REVIEW

**Comment by fverloop** [comment]
I've removed colorguard, because it wasn't really useful.

The stylelint task now only runs when there's a configuration in `package.json` or a `.stylelintrc` file.

File: gulpfile.coffee [hunk 1]
```diff
@@ -161,13 +165,33 @@ gulp.task "pages", ->
...
+	if settings.stylelint or fs.existsSync(projectPath("", ".stylelintrc"))
+		gulp.src(projectPath(paths.scss, "**/*.scss"))
+			.pipe(plumber())
+			.pipe(stylelint({
+				reporters: [
+					{formatter: 'string', console: true}
+				]
+		}))
+
+
```
Comment (koenbok) [inline gulpfile.coffee]: I think you have to pass back an emptytask if nothing needs to happen.

File: gulpfile.coffee [hunk 2]
```diff
@@ -300,19 +324,23 @@ gulp.task "server", (cb) ->
...
+		"applet", "blockquote", "abbr", "acronym", "cite", "del", "dfn", "kbd", "samp", "strike", "sup", "tt", "dt", "fieldset", "legend", "caption", "tfoot", "thead", "th", "figcaption", "hgroup", "mark", "blockquote", "blockquote:after", "blockquote:before", "textarea:focus"
+	]
+
+	gutil.log(gutil.colors.green("Unused CSS"))
 	return gulp.src(buildPath(paths.scss, "style.css"))
 		.pipe(purify(
-			[buildPath(paths.javascript, "**/*.js"), buildPath("", "**/*.html")],
-			{rejected: true}
+			[buildPath("", "**/*.html"), buildPath("", "**/*.js")],
+			{rejected: true, whitelist: commonResetClasses}
```
Comment (koenbok) [inline gulpfile.coffee]: This seems a bit project specific. Is there any way we can not hardcode these, and read them from some configuration file?

File: gulpfile.coffee (hunk 1 above)
Comment (koenbok) [inline gulpfile.coffee]: @fverloop `'` vs `"`

Comment (koenbok) [inline gulpfile.coffee]: YOU"RE BUILDING A LINTER FOR CHR*ST SAKE ;-)

**Comment by fverloop** [comment]
You can enable autoprefixer by adding it to the `style` property in `config.coffee`

image

File: gulpfile.coffee (hunk 1 above)
Comment (nvh) [inline gulpfile.coffee]: I think you meant YOU`'`RE instead of YOU`"`RE 😂

**Comment by koenbok** [comment]
LGTM

**Pull request merged**